### PR TITLE
Typescript conversion

### PR DESCRIPTION
**Pull request by kavsingh**
As per title

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -8,52 +8,58 @@
...
     "prepare": "npm run build",
     "predeploy": "cd example && npm install && npm run build",
     "deploy": "gh-pages -d example/build"
   },
+  "dependencies": {},
   "peerDependencies": {
-    "prop-types": "^15.5.4",
-    "react": "^15.0.0 || ^16.0.0",
-    "react-dom": "^15.0.0 || ^16.0.0"
+    "prop-types": ">=15.7.0",
```
Comment (kavsingh) [inline package.json]: Not sure if we should have this as a hard peer dep

File: src/index.tsx [hunk 2]
```diff
@@ -0,0 +1,63 @@
+import React, {
+  forwardRef,
+  ReactNode,
+  FunctionComponent,
+  createContext,
+  HTMLAttributes,
+  DetailedHTMLProps,
+  createElement,
+} from 'react'
+import PropTypes from 'prop-types'
```
Comment (kavsingh) [inline src/index.tsx]: As above, not sure if we should require that the end-user have `prop-types` installed

**Comment by kavsingh** [comment]
@osartun left some comments around `prop-types`, not sure if we should require that the end-user has it installed. Also we should probably bump the package version - major? minor?

File: src/index.tsx [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+export interface DocumentOutlineSectionContext {
```
Comment (osartun) [inline src/index.tsx]: Does this need to be `export`ed?
`level` and `title` would be read and set by the components in here exclusively, so I think there's no need that the outside world knows about these context properties or the context in general.

**Review by osartun — APPROVED** [review]
Wow! That's a substantial upgrade! 👏 
You ran the tests and everything works?

File: src/index.tsx (hunk 3 above)
Comment (kavsingh) [reply src/index.tsx]: that's a good point, will drop it

**Comment by kavsingh** [comment]
@osartun CI and local testing works so it looks ok?
Also, what about them prop-types?

**Comment by osartun** [comment]
@kavsingh 
> left some comments around prop-types, not sure if we should require that the end-user has it installed.

I guess now that TS is in place, prop-types is redundant. What do you think?

> Also we should probably bump the package version - major? minor?

Not an expert on this, but I'd say major as this is basically a rewrite. As it passes the tests it's no breaking changes, but, idk, I still think major makes sense.

> CI and local testing works so it looks ok?

I wasn't aware that I even configured a CI for this. Yeah, if this works then it looks ok! 👏

**Comment by osartun** [comment]
Different question:
```
"name": "react-document-section",
```
in `package.json` would need to change if we move it over to the 8fit namespace, right?

But maybe it makes sense to apply this change in a separate PR? As this isn't related to the TS conversion.

**Comment by kavsingh** [comment]
@osartun yeah i reckon we drop non-essential peer deps, so ciao prop-types :) i'd go for minor version bump tbh, seeing as semantic versioning is mostly about the public api, which is identical except for typescript support... is typescript support enough for a major? my guess is no. although, is dropping prop-types also enough for a major? my guess is also no

**Comment by osartun** [comment]
@kavsingh Okay, fair point. Then minor it is and drop prop-types! :)

**Pull request merged**